src/block_cache/block_mgr.c

```diff
@@ -60,7 +123,40 @@ static int
 __bm_checkpoint(
   WT_BM *bm, WT_SESSION_IMPL *session, WT_ITEM *buf, WT_CKPT *ckptbase, bool data_checksum)
 {
-    return (__wt_block_checkpoint(session, bm->block, buf, ckptbase, data_checksum));
+    WT_BLOCK *block, *tblock;
+    WT_CONNECTION_IMPL *conn;
+    bool found;
+
+    conn = S2C(session);
+    block = bm->block;
+
+    WT_RET(__wt_block_checkpoint(session, block, buf, ckptbase, data_checksum));
+
+    /*
+     * Close previous primary objects that are no longer being written, that is, ones where all
+     * in-flight writes have drained. We know all writes have drained when a subsequent checkpoint
+     * completes, and we know the metadata file is the last file to be checkpointed. After
+     * checkpointing the metadata file, review any previous primary objects, flushing writes and
+     * discarding the primary reference.
+     */
+    if (strcmp(WT_METAFILE, block->name) != 0)
+        return (0);
+    do {
+        found = false;
+        __wt_spin_lock(session, &conn->block_lock);
+        TAILQ_FOREACH (tblock, &conn->blockqh, q)
+            if (tblock->close_on_checkpoint) {
+                tblock->close_on_checkpoint = false;
+                __wt_spin_unlock(session, &conn->block_lock);
+                found = true;
+                WT_RET(__wt_fsync(session, tblock->fh, true));
+                WT_RET(__bm_close_block(session, tblock));
```

Comment:
@keitharnoldsmith is this the line you mean for WT-9503?

Reply:
I'll add a comment to that ticket.